### PR TITLE
Add ioctl

### PR DESCRIPTION
**Pull request by npmccallum**

### CODE REVIEW

File: sgx-types/src/attr.rs [hunk 1]
```diff
@@ -37,6 +37,12 @@ pub struct Attributes {
     xfrm: Xfrm,
 }
 
+impl Attributes {
+    pub const fn new(flags: Flags, xfrm: Xfrm) -> Self {
```
Comment (lkatalin) [inline sgx-types/src/attr.rs]: Just curious why this is `const`.

File: sgx-types/src/secs.rs [hunk 2]
```diff
@@ -55,7 +55,9 @@ testaso! {
 }
 
 impl Secs {
-    pub fn new(size: u64, base: u64, ssa: u32, contents: &Contents) -> Self {
+    pub const SIZE_MAX: u64 = 0x1_000_000_000;
```
Comment (lkatalin) [inline sgx-types/src/secs.rs]: This is a good addition.

File: sgx-traits/src/lib.rs [hunk 3]
```diff
@@ -15,40 +15,30 @@
...
+pub trait Builder: Sized {
+    type Enclave: Enclave;
 
-    fn new(secs: Secs) -> Result<Self>;
+    fn new(secs: secs::Secs) -> Result<Self>;
 
-    unsafe fn add_tcs(&mut self, tcs: Tcs, offset: usize) -> Result<Offset<Tcs>>;
-
-    unsafe fn add_struct<T>(
+    unsafe fn add(
```
Comment (lkatalin) [inline sgx-traits/src/lib.rs]: This seems to replace the functionality of `add_tcs`, `add_struct`, and `add_slice`, is that right?

File: sgx-sys/src/lib.rs [hunk 4]
```diff
@@ -114,8 +82,51 @@ mod test {
...
+        let mut off = 0;
+        for f in &flags {
+            #[repr(C, align(4096))]
+            struct Page([u8; 4096]);
+            let page = Page([0u8; 4096]);
+
+            // TCS pages MUST NOT specify permissions.
+            let si = page::SecInfo::new(page::Flags::empty(), page::Class::Tcs);
+            eprintln!("{:?} {:?} {:?}", off, *f, si);
+            assert!(unsafe { builder.add(off, &page.0, *f, si).is_ok() });
```
Comment (lkatalin) [inline sgx-sys/src/lib.rs]: Is this asserting that the builder was able to add a TCS page with the particular flag (`empty` for TCS) and secinfo combination at an offset that advances by one page each time?

File: sgx-types/src/attr.rs (hunk 1 above)
Comment (npmccallum) [reply sgx-types/src/attr.rs]: `const` functions can be evaluated at compile time. For example: https://doc.rust-lang.org/unstable-book/language-features/const-fn.html

File: sgx-traits/src/lib.rs (hunk 3 above)
Comment (npmccallum) [reply sgx-traits/src/lib.rs]: Correct.

File: sgx-sys/src/lib.rs (hunk 4 above)
Comment (npmccallum) [reply sgx-sys/src/lib.rs]: No. `TCS` flags have no permissions. Therefore, the set of flags is empty.

**Review by jarkkojs — APPROVED** [review]

**Pull request merged**